### PR TITLE
chore: Replace `reth-provider` dependency for `reth-eth-types` with direct dependencies

### PR DESCRIPTION
**Pull request by citizen-stig**
## Intro

I wanted to use `reth-eth-types` crate in Sovereign SDK rollup, in order to utilize known error types, such as `EthApiError` and `RpcInvalidTransactionError`, considering that they support conversion to jsonrpsee::ErrorObject.

When I tried to plug it into my workspace I've got weird error:

```rust
error[E0277]: the trait bound `usize: IteratorIndex<&mut StaticFileProviderOperationIter>` is not satisfied
   --> /Users/nikolai/.cargo/git/checkouts/reth-36d3ea1d1152b20c/4107df8/crates/storage/provider/src/providers/static_file/metrics.rs:107:51
    |
107 | #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, EnumIter)]
    |                                                   ^^^^^^^^ the trait `IteratorIndex<&mut StaticFileProviderOperationIter>` is not implemented for `usize`
    |
    = help: the following other types implement trait `IteratorIndex<I>`:
              RangeFrom<usize>
              RangeFull
              RangeInclusive<usize>
              RangeTo<usize>
              RangeToInclusive<usize>
              std::ops::Range<usize>
note: required by a bound in `itertools::Itertools::get`
   --> /Users/nikolai/.cargo/registry/src/index.crates.io-6f17d22bba15001f/itertools-0.13.0/src/lib.rs:554:12
    |
551 |     fn get<R>(self, index: R) -> R::Output
    |        --- required by a bound in this associated function
...
554 |         R: traits::IteratorIndex<Self>,
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `Itertools::get`
    = note: this error originates in the derive macro `EnumIter` (in Nightly builds, run with -Z macro-backtrace for more info)
```

But on closer inspection I've noticed that `reth-eth-types` depend on `reth-provider`, which brings a lot of dependencies into the tree, which won't be needed.

I decided to address that by using types directly and not via provider.

## Summary of the changes

* Use `reth-chain-state` and `reth-storage-api` directly instead of `reth-provider` in `reth-eth-types` and in `reth-transaction-pool`
* Move ` ChainSpecProvider` trait to `reth-chainspec` and re-export it in `reth-provider`

### CODE REVIEW

File: crates/storage/provider/src/traits/spec.rs
Comment (mattsse) [inline crates/storage/provider/src/traits/spec.rs:1]: I think we can delete this file and re-export from traits.mod instead

File: crates/chainspec/src/spec.rs [hunk 1]
```diff
@@ -808,6 +808,13 @@ impl From<Genesis> for ChainSpec {
     }
 }
 
+/// A trait for reading the current [`ChainSpec`].
+#[auto_impl::auto_impl(&, Arc)]
+pub trait ChainSpecProvider: Send + Sync {
+    /// Get an [`Arc`] to the [`ChainSpec`].
+    fn chain_spec(&self) -> Arc<ChainSpec>;
+}
```
Comment (mattsse) [inline crates/chainspec/src/spec.rs:816]: this is reasonable

**Review by mattsse — CHANGES_REQUESTED** [review]
cool, only have a nit.

I like this a lot, there are a few other structural improvements wip.

File: crates/storage/provider/src/traits/spec.rs
Comment (citizen-stig) [reply crates/storage/provider/src/traits/spec.rs:1]: Done!

**Review by mattsse — APPROVED** [review]

**Comment by citizen-stig** [comment]
@mattsse , all checks have passed now, thak you for a quick review!

**Pull request merged**